kirby/modules.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+namespace Kirby;
+
+use A;
+use F;
+use Dir;
+use Response;
+
+class Modules {
+
+  // core
+  public $controllers = array();
+  public $languages   = array();
+  public $models      = array();
+  public $snippets    = array();
+  public $tags        = array();
+  public $templates   = array();
+  public $autocss     = array();
+  public $autojs      = array();
+
+  // panel
+  public $blueprints  = array();
+  public $fields      = array();
+
+
+  /**
+   * Magic getter for module arrays
+   */
+  public function __call($method, $arguments) {
+    return isset($this->{$method}) ? $this->{$method} : null;
+  }
+
+  /**
+   * Registers a plugin directory for a module type
+   */
+  public function register($module, $dir) {
+    if(!isset($this->{$module})) return false;
+
+    // prepare for auto-assets
+    if($module === 'autocss' or $module === 'autojs') {
+      $route = 'assets/modules/' . sha1($dir);
+      $dir   = array($dir, $route);
+      $this->assets($route, $dir);
+    }
+
+    array_push($this->{$module}, $dir);
+    return true;
+  }
+
+  /**
+   * Creates route for plugin assets
+   */
+   public function assets($route, $root) {
+    if($page = page($route)) return false;
+
+    return kirby()->routes(array(
+      array(
+        'pattern' => trim($route, '/') . '/(:all)',
+        'action'  => function($file) use($route, $root) {
+          $path = $root . DS . $file;
+          if(is_file($path)) {
+            $extension = substr(strchr($path, '.'), 1);
+            return new Response(f::read($path), $extension);
+          } else {
+            return site()->errorPage();
+          }
+        }
+      )
+    ));
+  }
+
+
+  public function getAsset($module, $file) {
+
+    if($dirs = $this->{$module}()) {
+      foreach($dirs as $dir) {
+        $root = $dir[0] . DS . $file;
+        $url  = $dir[1] . DS . $file;
+        if(f::exists($root)) return $url;
+      }
+    }
+
+    return null;
+
+  }
+
+  public function getFile($module, $file, $ext, $default = null) {
+
+    if($dirs = $this->{$module}()) {
+      // $ext as array without dots
+      $ext = array_map(function($e) {
+        return ltrim($e, '.');
+      }, (array)$ext);
+
+      // add default location as possible source
+      if(!is_null($default)) array_unshift($dirs, $default);
```

Comment:
Couldn't this method use `allRoots()` to keep it DRY? This line would then be added to the `allRoots()` method.